audiostream/src/ni/media/audio/source/gstreamer_file_source.cpp

```diff
@@ -0,0 +1,337 @@
+    GstElement* sink      = gst_element_factory_make( "appsink", "sink" );
+
+    gst_bin_add_many( GST_BIN( m_pipeline.get() ), source, decodebin, queue, sink, nullptr );
+    gst_element_link_many( source, decodebin, NULL );
+    gst_element_link_many( queue, sink, NULL );
+
+    g_object_set( source, "location", path.c_str(), NULL );
+
+    g_signal_connect( decodebin, "pad-added", G_CALLBACK( &onPadAdded ), queue );
+
+    preroll_pipeline();
+    return sink;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::preroll_pipeline()
+{
+    gst_element_set_state( m_pipeline.get(), GST_STATE_PAUSED );
+
+    if ( wait_for_async_operation() != GST_STATE_PAUSED )
+        throw std::runtime_error( "gstreamer_file_source: pipeline doesn't preroll into paused state" );
+
+    gst_element_set_state( m_pipeline.get(), GST_STATE_PLAYING );
+
+    if ( wait_for_async_operation() != GST_STATE_PLAYING )
+        throw std::runtime_error( "gstreamer_file_source: pipeline doesn't preroll into playing state" );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::fill_format_info( GstStructure*                        caps_struct,
+                                              audio::ifstream_info::container_type container )
+{
+    m_info.container( container );
+    m_info.codec( audio::ifstream_info::codec_type::mp3 );
+    m_info.lossless( false );
+
+    gint64 num_frames = 0;
+    if ( !gst_element_query_duration( m_pipeline.get(), GST_FORMAT_DEFAULT, &num_frames ) )
+        throw std::runtime_error( "gstreamer_file_source: could not query duration from gstreamer" );
+
+    m_info.num_frames( num_frames );
+
+    int sample_rate = 0;
+    if ( !gst_structure_get_int( caps_struct, "rate", &sample_rate ) )
+        throw std::runtime_error( "gstreamer_file_source: could not query sample rate from gstreamer" );
+
+    m_info.sample_rate( sample_rate );
+
+    int num_channels = 0;
+    if ( !gst_structure_get_int( caps_struct, "channels", &num_channels ) )
+        throw std::runtime_error( "gstreamer_file_source: could not query number of channels from gstreamer" );
+
+    m_info.num_channels( num_channels );
+
+    m_info.format( create_runtime_format( caps_struct ) );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+pcm::runtime_format gstreamer_file_source::create_runtime_format( GstStructure* caps_struct )
+{
+    if ( auto format = gst_structure_get_string( caps_struct, "format" ) )
+    {
+        pcm::number_type number_type = gst_format_char_to_number_type( format[0] );
+        auto             srcDepth    = std::atol( format + 1 );
+        auto             endian      = ( strcmp( format, "BE" ) == 0 ) ? pcm::big_endian : pcm::little_endian;
```

Comment:
You probably want to use `GstAudioInfo` here and in the function above. it gives you a simple C struct with all these information you want to extract from the caps. Also check the `GstAudioFormatInfo` that is part of the `GstAudioInfo`.